### ISSUE TITLE
New Feature - Quick Picks Register

### ISSUE THREAD

**Issue by bullmike** [opensourcepos/opensourcepos#2164]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [] Reporting an issue of an unmodified OSPOS installation
- [x] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [x] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [x] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [x] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [x] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [x] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.2.3
- OSPOS git commit hash is: 
- PHP version is: PHP 7.0.32 
- MySQL or MariaDB version is: 
- OS and version is: 
- WebServer is: 
- Selected language is: 
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 

### Issue / Bug  / Question / New Feature

NEW FEATURE

Quick Picks for Register. 
I found it was a bit cumbersome to add items to the register so i added a quick pick option with a few lines of code. Once again, this would likely be accompanied by some configuration items.

application/controllers/Sales.php

     // get all items with image. could be modified to grab only items that are configured to show in quick 
         picks. In this situation, we want to return all of them as the item base is small.

     public function item_list()
        {
              $suggestions = array();

              foreach($this->Item->get_all(1)->result() as $row){
                       $suggestions[] = array('value' => $row->item_id, 'label' => $this->Item->get_search_suggestion_label($row), 'image'=> $row->pic_filename);
              }

              $suggestions = $this->xss_clean($suggestions);
               echo json_encode($suggestions);
        }

application/views/sales/register.php

Following code is placed in the $(document).ready(function() in the <script> tags at the bottom of the file. Could be cleaned up a bit.

        // Get Items
        $.getJSON( '<?php echo site_url($controller_name."/item_list"); ?>', function(data) {
                var items = [];
                $.each( data, function (key, val) {
                        var img;
                        if ( val.image ) {
                                img = val.image;
                        } else {
                                img = 'noimage.png';
                        }

                        items.push ( "<div style='float: left; margin: 10px; text-align: center;' class='clearfix col-xs-2'><div>" );
                        items.push ( "<img onclick='additem(" + val['value'] + ")' style='width: 30px; height:30px;' src='<?php echo base_url("uploads/item_pics/"); ?>" + img + "' />" );
                        items.push( "<br>" +val['label'] + "</div></div>" );
                });

                $("#quickpick").html(items.join(""));
        });

JS function to add the item when its clicked. Goes below the .ready() function

function additem(val)
{
        $('#item').val(val);
        $("#add_item_form").submit();
}

Bit of HTML to hold the items:

        <?php echo form_open($controller_name."/add", array('id'=>'add_item_form', 'class'=>'form-horizontal panel panel-default')); ?>
                <div class="panel-body form-group">
                        <ul>
                                <li class="pull-left first_li">
                                        <label for="item" class='control-label'><?php echo $this->lang->line('sales_find_or_scan_item_or_receipt'); ?></label>
                                </li>
                                <li class="pull-left">
                                        <?php echo form_input(array('name'=>'item', 'id'=>'item', 'class'=>'form-control input-sm', 'size'=>'50', 'tabindex'=>++$tabindex)); ?>
                                        <span class="ui-helper-hidden-accessible" role="status"></span>
                                </li>
                                <li class="pull-right">
                                        <button id='new_item_button' class='btn btn-info btn-sm pull-right modal-dlg' data-btn-new='<?php echo $this->lang->line('common_new') ?>' data-btn-submit='<?php echo $this->lang->line('common_submit')?>' data-href='<?php echo site_url("items/view"); ?>'
                                                        title='<?php echo $this->lang->line($controller_name . '_new_item'); ?>'>
                                                <span class="glyphicon glyphicon-tag">&nbsp</span><?php echo $this->lang->line($controller_name. '_new_item'); ?>
                                        </button>
                                </li>
                        </ul>
                </div>
                 <-- Container for QUICKPICK Items -->
                <div id='quickpick' class="panel-body form-group">

                </div>
                <div class='panel-body form-group'>
                        <ul>
                        </ul>
                </div>
        <?php echo form_close(); ?>

Looks something like this:

https://imagebin.ca/v/4H1oh63YNSSj

Please write your issue here. If a bug, please make sure to provide as much information as possible including configuration settings (e.g. Decimals set, Tax mode), language and steps to reproduce the bug.

**Comment by odiea** [comment]
How does the Register look when you have a 1000 + items?

**Comment by WebShells** [comment]
I was typing that @odeia...

**Comment by WebShells** [comment]
Maybe should consider pagination or categorizing items...

**Comment by WebShells** [comment]
Hello @bullmike, due to several needs for instant item pickup, i applied the following listing items depending on category and on custom field / and deleted = 0 ( if its set to YES, item will be displayed )

Maybe it will help you go forward with what you reached.

![grid](https://user-images.githubusercontent.com/26513147/46090561-a8ce4100-c1b9-11e8-9ab5-917741ad394b.png)

**Comment by bullmike** [comment]
Oh I realize that this isn't going to support 1000+ items.  Fortunately, that won't be an issue on my end. This was quick add-in as the functionality wasn't currently there.  

Thanks for the tip WebShells, I had the same idea to group based on category and for pagination, just didn't really need it at the moment. I was hoping this type of feature would be included in a future release and wanted to mention for the developers.

I've also added a few other items to support some requirements on our end.  If you are interested, they are as follows:

ITEM CHANGES:

Item type 'Calculated value'  with sub types: Field Calculation, Cart Calculation and Rate Calculation

Field Calculation:

If an item is defined as a 'Field Calculation' type, you can specify a formula in the ITEM definition. The register will add a field where you can type in a value. On change, the associated formula is applied and the price is updated with the result. In our situation we are using this to add a 'Insurance' line item where a customer may request to insure their 'shipment' for a $ figure.

Cart Calculation:

If an item is defined as a 'Cart Calculation' type, you can specify a formula in the ITEM definition. The formula is applied to the carts total and the associated price field is updated with the result. this allows us to apply a fee to the total of the cart ( like a tax ), except it is reflected as a line item. We can use this to apply a refrigeration or heating fee to the subtotal of the cart. I included a flag on the ITEM definition so you may exclude certain items from the calculation. Now that i write this, it would probably make more sense to change the flag to 'Include in Cart Calculation'.

Rate Calculation:

If an item is defined as a 'Rate Calculation' type, you may select a Rate Code from a drop down (linked to a new module called Rates). On the register, a field is displayed that allows for the entering of a 'Weight' value. when the value is changed, the Rate module is called and the subtotal is calculated based off the values defined in the Rate module. This allows for dynamic subtotals based on the 'weight' of an item.

SALES LOCATION
Added a drop down on the login screen that allows the employee to select their location. The location is essentially a token to assist with the reporting features. Eg..  Payment Summary by location,  Transaction Reports, etc.. I found this functionality is mandatory if using the system at more than one location as it allows you to independently report on sales data. Perhaps I'm missing something that is included in the core functionality?

PAYMENT TYPES:
Added sub types for the Major Card Types -  reporting purposes.
Added an 'ON Account' Type and 'Collect' Type. Collect is sort of an IOU as the payment will be completed at the remote end.


Upcoming changes:
1. Add functionality to allow the specification of 'People' for Pickup and Delivery Addresses at checkout.


Cheers,
Mike

**Comment by engehab87** [comment]
Hello Guys, any news on the Code, as this feature will really help alot in fast transactions for the Cashier.

**Comment by andriux1990** [comment]
> ![grid](https://user-images.githubusercontent.com/26513147/46090561-a8ce4100-c1b9-11e8-9ab5-917741ad394b.png)

Hello @WebShells!!
what tube to do to sort them by category?

**Comment by srikantamishra** [comment]
> 
> 
> Hello @bullmike, due to several needs for instant item pickup, i applied the following listing items depending on category and on custom field / and deleted = 0 ( if its set to YES, item will be displayed )
> 
> Maybe it will help you go forward with what you reached.
> 
> ![grid](https://user-images.githubusercontent.com/26513147/46090561-a8ce4100-c1b9-11e8-9ab5-917741ad394b.png)

Can u please post the code for others. i need the same.

**Comment by WebShells** [comment]
Hello @srikantamishra, the code above isn't written in a way that meets CI standards, thus if applied your version won't be supporting future releases and might end  messed up. Preferably we should discuss adding it as a feature to OSPOS... @daN4cat @jekkos @SteveIreland ?

**Comment by SteveIreland** [comment]
I'll have to pass on this one ... I'm committing to multi-currency sales and purchasing.

**Comment by engehab87** [comment]
@bullmike please share the current code for me to continue after it and not start from scratch

**Comment by bullmike** [comment]
@engehab87 the code is already posted in the thread. If you want to add grouping you would add the item category to the function in the sales controller and use some css or javascript in the register screen. You should be able to just paste that code into your system and have it work without grouping but you will be limited in terms of the how many items you can display on the screen.

**Comment by bullmike** [comment]
Also, the code above should work through upgrades, however, you will have to integrate it once again.

**Comment by andriux1990** [comment]
Hello @bullmike!! I did what you said to show the categories

![captura de pantalla de 2018-11-29 13-45-28](https://user-images.githubusercontent.com/29256343/49248753-124c1500-f3e0-11e8-8d17-7aa2dde0bb53.png)
![captura de pantalla de 2018-11-29 13-45-57](https://user-images.githubusercontent.com/29256343/49248825-3d366900-f3e0-11e8-9ba3-f19ec90a609f.png)
![captura de pantalla de 2018-11-29 13-46-12](https://user-images.githubusercontent.com/29256343/49248857-50493900-f3e0-11e8-90ea-2f17ac5fb1a4.png)

Obviously now there are only categories without the elements within them.
How to do it so that when I click on a category, show me all the types of that category?

**Comment by bullmike** [comment]
Good Question. 

You could make sure of the Bootstrap 3 button drop downs:

format:

 <div class="btn-group">
  <button type="button" class="btn btn-default dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
    Action <span class="caret"></span>
  </button>
  <ul class="dropdown-menu">
    <li><a href="#">Action</a></li>
    <li><a href="#">Another action</a></li>
    <li><a href="#">Something else here</a></li>
    <li><a href="#">Separated link</a></li>
  </ul>
</div>

 
Return a json object:

$array = array( 'category1' => array('item_id_1' => 'item_id_label', 'item_id_2' => 'item_id_2_label') );
json_encode($array);
{ 'category_name' :  { 'item_id_1' => 'item_id_label' ,  'item_id_2' : 'item_id_2_label'   } }

var items = [];
$.each ( data, function ( key , val) {
        items.push( "<div class='btn-group'>");
        items.push( " <button type='button' class='btn btn-default dropdown-toggle' data-toggle='dropdown' aria-haspopup='true' aria-expanded='false'> " );
        items.push( key + " <span class='caret'></span>");
        items.push(" </button>" );
        items.push(" <ul class='dropdown-menu'> " );
        $.each(val, function ( ikey, ival ) {
            items.push(" <li><a onclick='additem(" + ikey + ")'> + ival + "</a></li>"); 
        });
        items.push( " </ul> ");
        items.push( "</div> " );
    
});
$("quickpick").html(items.join(""));

i just free handed this into the comment field, so it may need debugging.

**Comment by andriux1990** [comment]
thank you very much for your help @bullmike now I'm just trying to see where all that code goes..

![screenshot_2018-12-01 https pastebin com](https://user-images.githubusercontent.com/29256343/49325387-04dd7a80-f507-11e8-86d4-4f0019318898.png)
![screenshot_2018-12-01 new feature - quick picks register issue 2164 opensourcepos opensourcepos](https://user-images.githubusercontent.com/29256343/49325388-132b9680-f507-11e8-9bf3-9c6a39db4c87.png)

I was doing some tests and it did not work, I think I copied the code where it should not be.
Thank you for your help in this way of learning for me. If you could help me achieve it, it would make me happy. I would really appreciate it

**Comment by bullmike** [comment]
Hi,

Here you go:

https://pastebin.com/raw/J2dKGVzn

**Comment by daN4cat** [comment]
Would be nice to have alternati e Register submitted as contribution and selectable from configuration. What do you think guys?

**Comment by jekkos** [comment]
@daN4cat I would then suggest to move the dinners table functionality to a separate register 'view' instead of having it configurable as a config option in ospos. This is already a good example of such an 'alternative' register focusing on a specific type of business.

**Comment by daN4cat** [comment]
Agreed. It would be better to have multiple specialised Registers and a config param to select which one you want.
Similar to Themes and Receipts.

Ideally we should tidy up the APIs and move to REST based one in parallel...

**Comment by bullmike** [comment]
@andriux1990  - not sure. I didn't even know there was a version 3.3.0 . where did you get version 3.3.0 ?

**Comment by andriux1990** [comment]
I think the problem is here in the category.
$categories[$result->**category**][] = $result->name;
In version 3.3.0 you can not filter category, and I do not know why?

public function get_item_categories()
    {
        $this->load->model('item');        
        $items = $this->Item->get_all();
       // print_r($items->result());
        $categories = array();        
        
        foreach ( $items->result() as $result) {
            $categories[$result->category][] = $result->name;
        }
        
        echo json_encode($categories);
    }
![screenshot_2018-12-08 open source point of sale powered by ospos 3 3 0](https://user-images.githubusercontent.com/29256343/49688414-1d393080-fad7-11e8-915a-65035e7023b8.png)

**Comment by odiea** [comment]
If the code was 100% complete for further testing that would be great.

**Comment by WebShells** [comment]
Actually our gird is still the most fitting one so far, but am sure the code is somehow messy so  if you like ill try to rework on it asap and share... It lists the categories that has item with grid custom field = yes, now should be changed to attributes. It displays divided into categories each has items under it, might need pagination too and back office checkbox to enable and disable feature...

**Comment by engehab87** [comment]
any progress ????

**Comment by engehab87** [comment]
@odiea the problem when selecting from the droplist is that it retrieves the item id less than the actual by "1", ex. if the item id is 3 in the database then it retrieves item id 2.

any suggestions ?

**Comment by odiea** [comment]
The following is not sanctioned by OSPOS so use at your own risk. It is using a split button for the dropdown. Select the Item number or id to add to a sale. Clicking on the name just closes the button. Good luck and hope this helps someone out. Follow the directions in the included cat_buttons .text.
![2019-01-09 05_48_10-window](https://user-images.githubusercontent.com/10962177/50900629-0423e800-13d3-11e9-8220-daa76a388376.png)


[cat_buttons.txt](https://github.com/opensourcepos/opensourcepos/files/2740925/cat_buttons.txt)

**Comment by engehab87** [comment]
@odiea i was able to resolve it after Replacing public function get_categories() with public function get_category() in both Models/item and controllers/Sales

the point is that the dropdown list states both the item_id and the name in the list which the list should be stating only the name and the name to reflect the item_id.

any help in that ?

**Comment by odiea** [comment]
Sorry This is all I could do  to make it work.

**Comment by andriux1990** [comment]
@engehab8, What do you mean exactly

**Comment by odiea** [comment]
When clicking the id the item submits  ok but clicking on the name it does not submit it. I think he would like to click on the Name and have it submitted.

**Comment by engehab87** [comment]
@odiea exactly what i'm looking for and adding to that it will be great to have the items list as grid under each category

**Comment by FAB123** [comment]
> Hello @bullmike, due to several needs for instant item pickup, i applied the following listing items depending on category and on custom field / and deleted = 0 ( if its set to YES, item will be displayed )
> 
> Maybe it will help you go forward with what you reached.
> 
> ![grid](https://user-images.githubusercontent.com/26513147/46090561-a8ce4100-c1b9-11e8-9ab5-917741ad394b.png)

nice, can you share code?

**Comment by andriux1990** [comment]
Any progress here?

**Comment by andriux1990** [comment]
![Screenshot_2020-01-11 Open Source Point Of Sale Powered by NMP 3 4 0](https://user-images.githubusercontent.com/29256343/72214461-1940c480-34c8-11ea-954a-1b4151ade52d.png)

**Comment by WebShells** [comment]
@andriux1990 is this version based on a version available on codecanyon? 🤔
Ive seen someone working on it before....

**Comment by andriux1990** [comment]
Codecanyon is messy and has no style.
@WebShells not really. in style and css has more things similar to phppos

**Comment by WebShells** [comment]
Yes, that's clear..

**Comment by andriux1990** [comment]
My header.php is similar to phppos and it took me about two months to finish it. use an awesome foun for icons and an admin-like interface

**Comment by ConceptFreak** [comment]
> > Hello @bullmike, due to several needs for instant item pickup, i applied the following listing items depending on category and on custom field / and deleted = 0 ( if its set to YES, item will be displayed )
> > Maybe it will help you go forward with what you reached.
> > ![grid](https://user-images.githubusercontent.com/26513147/46090561-a8ce4100-c1b9-11e8-9ab5-917741ad394b.png)
> 
> nice, can you share code?

Can you provide the code for the work you did

**Comment by andriux1990** [comment]
Is it true @WebShells  because you don't do it? your code looks interesting

**Comment by ConceptFreak** [comment]
I am working on it if i get something nice i will share the code.
It would be very helpful if someone who already has done it shares the code.

**Comment by WebShells** [comment]
A year ago... i did it for someone but i didn't apply any updates to newer releases and it's based on custom fields, just got it working as is without code standardizing...
We got some minor fixes for the upcoming release... Will try to pass by this one or we will postpone to next milestone...

**Comment by ConceptFreak** [comment]
I have done it here is the code
 
Model CODE  application\models\Item.php
Add it at the bottom of the model before last  }

public function get_categorieso()
{
$this->db->select('category');
$this->db->from('items');
$this->db->where('deleted', 0);
$this->db->distinct();
$this->db->order_by('category', 'asc');
$this->db->order_by('name', 'asc');
return $this->db->get();
}
public function get_cat_items($cat)
{
$this->db->select('item_id,name,pic_filename');
$this->db->from('items');
$this->db->where('deleted', 0);
$this->db->where('category', $cat);
return $this->db->get();
}




Controler CODE
application\controllers\Sales.php
add this at the end before }


public function get_item_categories()
    {
        $this->load->model('item');        
		$item = $this->Item->get_categorieso();     
      // $this->db->order_by('name', 'asc');
	   $categories = array();               
        foreach($item->result() as $result)
		{            
		    $categories[$result->category][]=$result->category; 
	    }	
		
        echo json_encode($categories);
    }
    
    public function get_cat_items()
    {
        $cat = $this->input->post('cat');
        $this->load->model('item');        
		$item = $this->Item->get_cat_items($cat);     
       
	    $categories = array();               
        foreach($item->result() as $result)
		{            
			 $categories[]=array($result->item_id,$result->name,$result->pic_filename); 	
	    }	
        echo json_encode($categories);
    }


View CODE
Add this at the end before </script>



$.getJSON( '<?php echo site_url($controller_name."/get_item_categories"); ?>', function(data)
 {
    var items = [];	
	items.push("<nav class='navbar navbar-expand-lg navbar-light bg-light'>");
    items.push("<a class='navbar-brand' href='#'>Categories</a>");
    items.push("<button class='navbar-toggler' type='button' data-toggle='collapse' data-target='#navbarNavAltMarkup' aria-controls='navbarNavAltMarkup' aria-expanded='false' aria-label='Toggle navigation'>");
    items.push("<span class='navbar-toggler-icon'></span>");
    items.push("</button>");
    items.push("<div class='collapse navbar-collapse' id='navbarNavAltMarkup'>");
    items.push("<div class='navbar-nav'>");   
        $.each( data, function (key,val)
		{     
	 items.push(" <a cat='"+key+"' class='catholder'>" + key + "</a>");          
    
        });

	  items.push("</div>");
      items.push("</div>");
      items.push("</nav>");
        
    $("#quickpick").html(items.join(""));
	$('#quickpickhide').dblclick(function() { $('#quickpick').toggle()} );
	
	 // Get Items   
	
});

$(document).on("click",".catholder", function () {
   var category = $(this).attr('cat'); // or var clickedBtnID = this.id
   var items = [];	
   $.ajax({
        url: "<?php echo site_url($controller_name.'/get_cat_items'); ?>",
        type: 'POST',
        data: {cat: category},
        dataType: 'json',
        success: function(res){
		console.log(res);
		items.push("<div class='row'>");	
        $.each(res,function(ikey,ival){
			items.push("<div class='col-sm-2 col-xs-4 category-item'>");
			items.push ( "<img class='img-responsive' onclick='additem(" + ival[0] + ")'  src='<?php echo base_url("uploads/item_pics/"); ?>" + ival[2]+ "' />" );
            items.push("<a onclick='additem(" + ival[0] + ")'>"+ival[1]+"</a>");
			items.push("</div>");
		}); 
			items.push("</div>");
		$("#itemcontainer").html(items.join(""));  
        }
    });
});


    
function additem(val)
{
$('#item').val(val);
$("#add_item_form").submit();
}
$('.dropdown-toggle').dropdown();




After that you can put this also inside the VIEW before the opening of <script>
this will show the content
<div id='quickpick' align='center'>            		   			
</div>
<div id='itemcontainer'>     		   			
</div>

**Comment by ConceptFreak** [comment]
![Capture](https://user-images.githubusercontent.com/46026370/73205091-14377280-4162-11ea-9dd9-02f081997692.PNG)

**Comment by odiea** [comment]
Can you show where and what code goes where to actually view this new list?

**Comment by andriux1990** [comment]
Ajax is also needed to add products without reloading the page and flies at the beginning each time a product is add

**Comment by andriux1990** [comment]
Looks good

**Comment by engehab87** [comment]
@ConceptFreak Can you show where and what code goes where to actually view this new list?

**Comment by mrbozkaya** [comment]
It's awesome Job.

How can I apply this for "Item Kit"

Thank you,

**Comment by Hallsie** [comment]
Wait, so is this in the actual project or is this still something people are adding in later?
